enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,173 @@
+
+## Proposal
+
+Our main goal is to start using Cloud Controller Manager in OpenShift 4 on OpenStack. So we are going to use the implementation available in upstream as a part of [cloud-provider-openstack](https://github.com/kubernetes/cloud-provider-openstack/tree/master/pkg/cloud-controller-manager) repo.
+
+To maintain the lifecycle of the CCM we want to implement a cluster operator, that will handle all administrative tasks: deploy, restore, upgrade, healthchecks, and so on. Later this operator can be reused to manage CCMs for other platforms (AWS, Azure, GCP and so on).
+
+### Action plan
+
+#### Build OpenStack CCM image by OCP automation
+
+To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
+The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Test Cloud Controller Manager manually
+
+When all required components are built, we can manually deploy OpenStack CCM and test how it works.
+
+Actions:
+
+- Manually install CCM's [daemonset](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml) on a working OCP cluster deployed on OpenStack.
+
+- Update configuration of `kube-apiserver`, `kube-controller-manager` and `kubelet` by replacing `--cloud-provider openstack` with `--cloud-provider external` and removing `--cloud-config` parameters.
+
+**Note:** Example of manual testing: https://asciinema.org/a/303399?speed=2
+
+#### Write Cluster Cloud Controller Manager Operator
+
+The operator should be able to create, configure and manage different CCMs (including one for OpenStack) in OpenShift.
+
+Actions:
+
+- Create a new repo in OpenShift’s github: https://github.com/openshift/cluster-cloud-controller-manager-operator
+
+- Implement the operator, using [library-go](https://github.com/openshift/library-go) primitives.
+
+#### Build the operator image by OCP automation
+
+To start using CCM operator in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The image should be automatically tested before it becomes available. Dockerfile should be a part of the operator's repo.
+
+Actions:
+
+- Configure CI operator to build CCM operator image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Integrate the solution with OpenShift
```

Comment:
Does this list need to include disabling the existing cloud controller that exists within kube-controller-manager? Or can they co-exist? How can we configure this so that it works across multiple platforms?

Reply:
Not sure about usage across multiple platforms, might be possible when the cloud code is fully contained in a separate controller binary, but setting the `--cloud-provider external` is already disabling internal `kcm` cloud controller.

Reply:
Now "in-tree cloud provider" is just a module in kubernetes codebase. to disable it and start using the cloud controller manager we just need to replace `--cloud-provider openstack` to `--cloud-provider external` for these componets.

Reply:
> Not sure about usage across multiple platforms

To clarify, I was thinking along the lines of, if provider A supports out of tree, it needs these settings to be `external`, but provider B doesn't support out of tree yet, so needs to have these settings be `provider b`, somehow the configuration for these things will need to be templated right? 

Is there not an existing cloud controller that runs within (https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/ search for `cloud-node-lifecycle`) that would need to be disabled as part of this migration? It was my understanding that this functionality is taken over by CCM?

Reply:
OpenStack is already using a [custom kubelet.service](https://github.com/openshift/machine-config-operator/blob/adda352ad43deb16e1e63c0457094c8451342c35/templates/master/01-master-kubelet/openstack/units/kubelet.service.yaml) file so switching only the openstack platform to the external cloud provider should be straightforward.

Reply:
Ok that's fine then, I still think my second point about kube-controller-manager is unaddressed though. Kube-controller-manager has a `cloud-provider` arg, this is not currently mentioned in this proposal, I would assume we need to make sure that is either disabled or set to external for each platform post migration?

Looks like they're doing something to do with the cloud provider in the kube-controller-manager operator so I think this will need to be updated right? Or is the CCM responsibility completely separate from what kube-controller-manager is doing in the cloud?
https://github.com/openshift/cluster-kube-controller-manager-operator/blob/d56255faaced793c07f57ded4cbb42aab4fab879/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85